### PR TITLE
feat(vmip): validate IP address range and improve IP address service

### PR DESCRIPTION
**Pull request by LopatinDmitr**
## Description

- Add handling case, when trying to create VirtualMachineIPAddress outside the range specified in the module settings
- Add reason "VirtualMachineIPAddressOutOfTheValidRange" for "Bound" condition
- Refactor ip_address_service  with the use of net.IPNet to netip.Addr 
- Fix frequent reconciles for an object with the 'Released' status
- Add write to events, when the selected VirtualMachineIPLease belongs to a different namespace

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

### CODE REVIEW

File: images/virtualization-artifact/pkg/controller/service/errors.go [hunk 1]
```diff
@@ -23,3 +23,9 @@ var (
 	ErrDefaultStorageClassNotFound = errors.New("default storage class not found")
 	ErrDataVolumeNotRunning        = errors.New("pvc import is not running")
 )
+
+var (
+	ErrInvalidIpAddress      = errors.New("Invalid IP address format")
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/service/errors.go:28]: Error string should not be capitalized

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go [hunk 2]
```diff
@@ -53,73 +54,72 @@ func NewIpAddressService(
 	}
 }
 
-func (s IpAddressService) IsAvailableAddress(address string, allocatedIPs common.AllocatedIPs) bool {
-	ip := net.ParseIP(address)
+func (s IpAddressService) IsAvailableAddress(address string, allocatedIPs common.AllocatedIPs) error {
+	ip, err := netip.ParseAddr(address)
+	if err != nil || !ip.IsValid() {
+		return ErrInvalidIpAddress
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/service/ip_address_service.go:60]: Do we really need var for this err?

File: images/virtualization-artifact/pkg/controller/service/errors.go [hunk 3]
```diff
@@ -23,3 +23,9 @@ var (
...
+	ErrIpAddressAlreadyExist = errors.New("IP address is already allocated")
+	ErrIpAddressOutOfRange   = errors.New("IP address is out of range")
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/service/errors.go:30]: How about adding a check to the webhook to warn user?

File: images/virtualization-artifact/pkg/controller/service/errors.go [hunk 4]
```diff
@@ -23,3 +23,9 @@ var (
 	ErrDefaultStorageClassNotFound = errors.New("default storage class not found")
 	ErrDataVolumeNotRunning        = errors.New("pvc import is not running")
 )
+
+var (
+	ErrInvalidIpAddress      = errors.New("Invalid IP address format")
+	ErrIpAddressAlreadyExist = errors.New("IP address is already allocated")
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/service/errors.go:29]: ErrIpAddressAlreadyExist -> ErrIPAddressAlreadyExists

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go [hunk 5]
```diff
@@ -29,18 +30,18 @@ import (
 
 type IpAddressService struct {
 	logger      logr.Logger
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/service/ip_address_service.go:32]: Do you need logger for this service?

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go [hunk 6]
```diff
@@ -29,18 +30,18 @@ import (
...
-	ParsedCIDRs []*net.IPNet
+	ParsedCIDRs []netip.Prefix
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/service/ip_address_service.go:33]: Why public?

**Review by Isteb4k — CHANGES_REQUESTED** [review]

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go (hunk 6 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/ip_address_service.go:33]: Sorry, don't understand.

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go (hunk 2 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/ip_address_service.go:60]: Sorry, I don't quite understand what the problem is.

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go (hunk 5 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/ip_address_service.go:32]: Indeed, we write logs only in the constructor. We can opt out of member.

File: images/virtualization-artifact/pkg/controller/service/errors.go (hunk 3 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/errors.go:30]: Maybe. It's probably worth discussing separately.

File: images/virtualization-artifact/pkg/controller/service/errors.go (hunk 1 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/errors.go:28]: Fixed.

File: images/virtualization-artifact/pkg/controller/service/errors.go (hunk 4 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/errors.go:29]: Fixed in several places.

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go (hunk 2 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/ip_address_service.go]: Sorted it out. Fixed.

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go (hunk 6 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/ip_address_service.go:32]: Sorted it out. Fixed.

File: images/virtualization-artifact/pkg/controller/service/errors.go (hunk 3 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/errors.go]: Based on the results of the discussion, I added warn user in the webhook.

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go [hunk 7]
```diff
@@ -28,98 +29,95 @@ import (
...
 			}
 		}
 		// out of range
-		return false
+		return ErrIPAddressOutOfRange
 	}
+
 	// already exists
-	return false
+	return ErrIPAddressAlreadyExist
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/service/ip_address_service.go:73]: Fail early: change `!ok` to `ok` and return ErrIPAddressAlreadyExist inside if, not after.

File: images/virtualization-artifact/pkg/controller/vmip/internal/iplease_handler.go [hunk 8]
```diff
@@ -122,11 +124,37 @@ func (h IPLeaseHandler) createNewLease(ctx context.Context, state state.VMIPStat
...
+			mgr.Update(conditionBound.Status(metav1.ConditionFalse).
+				Reason(vmipcondition.VirtualMachineIPAddressLeaseAlready).
+				Message(fmt.Sprintf("VirtualMachineIPAddressLease %s is bound to another VirtualMachineIP",
+					common.IpToLeaseName(vmipStatus.Address))).
+				Condition())
+			h.recorder.Event(vmip, corev1.EventTypeWarning, vmipcondition.VirtualMachineIPAddressLeaseAlready, msg)
+		}
+
+		vmipStatus.Conditions = mgr.Generate()
+		return reconcile.Result{}, err
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/vmip/internal/iplease_handler.go:157]: Is it OK to reconcile on ErrIPAddressOutOfRange or ErrIPAddressAlreadyExist errors? They seem a final state for resource and we can `return reconcile.Result{}, nil` after updating status.

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go [hunk 9]
```diff
@@ -28,98 +29,95 @@ import (
...
+func toIPNet(prefix netip.Prefix) *net.IPNet {
+	return &net.IPNet{
+		IP:   prefix.Masked().Addr().AsSlice(),
+		Mask: net.CIDRMask(prefix.Bits(), prefix.Addr().BitLen()),
 	}
 }
 
-func isFirstLastIP(ip net.IP, cidr *net.IPNet) (bool, error) {
-	size := int(k8snet.RangeSize(cidr))
+func isFirstLastIP(ip netip.Addr, cidr netip.Prefix) (bool, error) {
```
Comment (diafour) [inline images/virtualization-artifact/pkg/controller/service/ip_address_service.go:104]: It'll be great to add simple unit tests for isFirstLastIP and toIPNet and for service public methods.

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go (hunk 7 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/ip_address_service.go]: Fixed it.

File: images/virtualization-artifact/pkg/controller/service/ip_address_service.go (hunk 9 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/service/ip_address_service.go:104]: I added tests.

File: images/virtualization-artifact/pkg/controller/vmip/internal/iplease_handler.go (hunk 8 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/vmip/internal/iplease_handler.go:157]: Indeed, he did so. 
Further reconciles will lead to nothing.

File: images/virtualization-artifact/pkg/controller/vmip/vmip_webhook.go [hunk 10]
```diff
@@ -71,14 +78,22 @@ func (v *Validator) ValidateCreate(ctx context.Context, obj runtime.Object) (adm
 					allocatedLease.Spec.VirtualMachineIPAddressRef.Name != vmip.Name) {
 				return nil, fmt.Errorf("VirtualMachineIPAddress cannot be created: the IP address %s has already been allocated by VirtualMachineIPAddress/%s in ns/%s", ip, allocatedLease.Spec.VirtualMachineIPAddressRef.Name, allocatedLease.Spec.VirtualMachineIPAddressRef.Namespace)
 			}
+
+			err = v.ipService.IsAvailableAddress(vmip.Spec.StaticIP, allocatedIPs)
+			if err != nil {
+				if errors.Is(err, service.ErrIPAddressOutOfRange) {
+					warnings = append(warnings, fmt.Sprintf("The requested address %s is out of the valid range",
+						vmip.Spec.StaticIP))
+				}
```
Comment (Isteb4k) [inline images/virtualization-artifact/pkg/controller/vmip/vmip_webhook.go:87]: An error was lost. If it's not a warning, log it

**Review by Isteb4k — APPROVED** [review]

File: images/virtualization-artifact/pkg/controller/vmip/vmip_webhook.go (hunk 10 above)
Comment (LopatinDmitr) [reply images/virtualization-artifact/pkg/controller/vmip/vmip_webhook.go:87]: It is a warning.

**Pull request merged**